src/utils/mtev_hyperloglog.c

```diff
@@ -62,33 +63,37 @@ mtev_hyperloglog_add(mtev_hyperloglog_t *hll, const void *data, size_t len)
   hash = mtev_hash__hash(data, len, 0x6F61567A);
 
   /* pick initial index based on precision */
-  index = hash >> bitarea;
+  index = (hash >> bitarea);
 
   /* shift out the index bits from the hash */
   hash = hash << hll->bitcount | (1 << (hll->bitcount -1 ));
 
   /* count leading zeroes */
   int l = __builtin_clz(hash) + 1;
 
+  uint8_t y = hll->regs[index];
+
   /* save the register if we got bigger */
-  if (l > hll->regs[index]) {
+  if (l > y) {
     hll->regs[index] = l;
   }
 }
 
 static inline double 
-mtev_hyperloglog_estimate(mtev_hyperloglog_t *hll, int *zero_count)
+mtev_hyperloglog_estimate(mtev_hyperloglog_t *hll, int *zero_register_count)
 {
-  int c;
-  double m, sum;
+  uint64_t int_sum = 0;
+  double m, sum = 0;
+  int c = (1 << hll->bitcount);
 
-  c = (1 << hll->bitcount);
   m = mtev_hyperloglog_alpha(hll->bitcount) * c * c;
-  sum = 0;
 
-  for (int i=0; i < c; i++) {
-    sum += 1.0 / (1 << hll->regs[i]);
+  for (int i = 0; i < c; i++) {
+    uint8_t x = hll->regs[i];
+    *zero_register_count += x == 0;
+    int_sum += c / (1 << x);
```

Comment:
`c / (1 << x)` seems wrong if `bitcount` is low (like, imagine `bitcount == 4`, then `c == 16`, and this division would generally bias the sum to be too small). But the choice of numerator seems to factor out when calculating the `double sum`, so you could just use any constant? I haven't thought this through completely, but `1 << (sizeof(int) * 8 - hll->bitcount)` should work, and would be most accurate possible without losing precision, I think. Or failing that, just using `65536` as a numerator should be pretty good.